Add placement deletes to avoid warnings on VC++

hoelzl opened the pull request:
When compiling with VC++ 2017, Godot generates huge nubmers of C4291 warnings about missing placement delete.

I have not found a way to disable these warnings using compiler options: AFAICT there is no equivalent to `-f-no-exceptions` for VC++ (there is only /EH to change the exception-handling model,
/GX is deprecated) and adding /wd4291  to the `disable_nonessential_warnings` list in the `SConstruct` file or even compiling with `warnings=no` does not disable the messages.

Placement delete is only called when placement new throws an exception, since Godot doesn't use exceptions this change should have no runtime effect.

Fixes #12654 (probably, difficult to say without log)

hoelzl committed:
Add placement deletes to avoid warnings on VC++

When compiling with VC++ 2017, Godot generates huge numbers of
C4291 warnings about missing placement delete.

I have not found a way to disable these warnings using compiler
options: AFAICT there is no equivalent to `-f-no-exceptions` for
VC++ (there is only /EH to change the exception-handling model,
/GX is deprecated) and adding /wd4291  to the
`disable_nonessential_warnings` list in the `SConstruct` file
or even compiling with `warnings=no` does not disable the
messages.

Placement delete is only called when placement new throws an
exception, since Godot doesn't use exceptions this change should
have no runtime effect.

Fixes #12654 (probably, difficult to say without log)

RandomShaper commented:
It'd be better to have them with their empty body in the header and remove them from the .cpp.

Also a comment about their purpose and wrapping them inside an #if restricting them to MSVC would be nice.

Nice contribution. That flood of undisableable warnings is annoying!

hoelzl commented:
I've added a comment and wrapped the code in `#ifdef _MSC_VER`.
I've also added `ERR_PRINT` statements to the implementations to make it clearer that these definitions are not meant to be used.
I didn't move the implementation to the header file. Moving them would inline the definitions and then lead to linker warnings, since the header is included in multiple libraries.

RandomShaper commented:
Perfect!

neikeq commented:
IMO better use `ERR_EXPLAIN` and ` CRASH_NOW`.

hoelzl commented:
Done, except that I've used `ERR_EXPLAINC` instead of `ERR_EXPLAIN` to avoid using `String`.
